### ISSUE TITLE
Likely false positive in diesel

### ISSUE THREAD

**Issue by weiznich** [obi1kenobi/cargo-semver-checks#1265]
### Which lint or lints are the issue

trait_added_supertrait

### Known issues that might be causing this

- [x] Is the flagged item defined in another crate, or a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))

### Steps to reproduce the bug with the above code

```
git clone https://github.com/diesel-rs/diesel
cd diesel/diesel
git checkout 8895eac
cargo semver-checks --only-explicit-features --baseline-version 2.2.10 --features "sqlite" --features "postgres" --features "mysql" --features "extras" --release-type minor
```

### Actual Behaviour

```
Description:
A non-sealed trait added one or more supertraits, which breaks downstream implementations of the trait
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#generic-bounds-tighten
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_added_supertrait.ron

Failed in:
  trait diesel::connection::Instrumentation gained Downcast in file /home/weiznich/Documents/rust/diesel/diesel/src/connection/instrumentation.rs:246
```

### Expected Behaviour

The relevant change in diesel is here: https://github.com/diesel-rs/diesel/pull/4129

Now the lint is correct that this adds a new super trait `Downcast`. As far as I understood your talk and #638 this is kind of expected as cargo semver-checks doesn't know anything about the `Downcast` trait yet and also it doesn't do any trait solving yet.

Now I still fill this issue as I feel this particular case has another rather obscure edge case.

The original definition of the instrumentation trait was:

```
pub trait Instrumentation: Send + 'static { /**/ } 
```

The PR added a super bound on the `Downcast` trait: https://docs.rs/downcast-rs/latest/downcast_rs/trait.Downcast.html, which in turn is implemented by any type implementing `Any`. Given that `Any` is a compiler built-in trait that's automatically implemented for any type that is `'static` this holds true for any type that previously implemented `Instrumentation` due to the existing bound. User defined unsized types do not exist in stable rust yet, so down stream users shouldn't be able to have them to implement the trait there.  Overall this shouldn't make this a breaking change, which in turn means this lint would be a false positive even if it had all the information that are currently missing.

### Verbose Lint Output

cargo semver-checks --only-explicit-features --baseline-version 2.2.10 --features "sqlite" --features "postgres" --features "mysql" --features "extras" --release-type minor --verbose                          
    Updating crates.io index
     Locking 144 packages to latest compatible versions
    Building diesel v2.2.4 (current)
 Documenting diesel v2.2.4 (/home/weiznich/Documents/rust/diesel/diesel)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 6.15s
       Built [   6.289s] (current)
     Parsing diesel v2.2.4 (current)
      Parsed [   0.318s] (current)
     Parsing diesel v2.2.10 (baseline, cached)
      Parsed [   0.206s] (baseline)
    Checking diesel v2.2.10 -> v2.2.4 (assume minor change)
    Starting 128 checks, 36 unnecessary on 20 threads
        PASS [   0.007s]       major        attribute_proc_macro_missing
        PASS [   0.112s]       major        auto_trait_impl_removed
        PASS [   0.024s]       major        constructible_struct_adds_field
        PASS [   0.035s]       major        constructible_struct_adds_private_field
        PASS [   0.014s]       major        constructible_struct_changed_type
        PASS [   0.012s]       major        declarative_macro_missing
        PASS [   0.011s]       major        derive_helper_attr_removed
        PASS [   0.013s]       major        derive_proc_macro_missing
        PASS [   0.132s]       major        derive_trait_impl_removed
        PASS [   0.011s]       major        enum_discriminants_undefined_non_exhaustive_variant
        PASS [   0.011s]       major        enum_discriminants_undefined_non_unit_variant
        PASS [   0.021s]       major        enum_marked_non_exhaustive
        PASS [   0.014s]       major        enum_missing
        FAIL [   0.018s]       major        enum_no_repr_variant_discriminant_changed
        PASS [   0.013s]       major        enum_non_exhaustive_tuple_variant_changed_kind
        PASS [   0.012s]       major        enum_now_doc_hidden
        PASS [   0.007s]       major        enum_repr_int_changed
        PASS [   0.011s]       major        enum_repr_int_removed
        PASS [   0.011s]       major        enum_repr_transparent_removed
        PASS [   0.013s]       major        enum_repr_variant_discriminant_changed
        PASS [   0.012s]       major        enum_struct_variant_changed_kind
        PASS [   0.023s]       major        enum_struct_variant_field_added
        PASS [   0.014s]       major        enum_struct_variant_field_missing
        PASS [   0.014s]       major        enum_struct_variant_field_now_doc_hidden
        PASS [   0.009s]       major        enum_tuple_variant_changed_kind
        PASS [   0.022s]       major        enum_tuple_variant_field_added
        PASS [   0.014s]       major        enum_tuple_variant_field_missing
        PASS [   0.009s]       major        enum_tuple_variant_field_now_doc_hidden
        PASS [   0.014s]       major        enum_unit_variant_changed_kind
        PASS [   0.021s]       major        enum_variant_added
        PASS [   0.039s]       major        enum_variant_marked_non_exhaustive
        PASS [   0.028s]       major        enum_variant_missing
        PASS [   0.007s]       major        exported_function_changed_abi
        PASS [   0.001s]       major        feature_missing
        PASS [   0.001s]       major        feature_not_enabled_by_default
        PASS [   0.020s]       major        function_abi_no_longer_unwind
        PASS [   0.025s]       major        function_abi_now_unwind
        PASS [   0.013s]       major        function_changed_abi
        PASS [   0.014s]       major        function_const_removed
        PASS [   0.011s]       major        function_export_name_changed
        PASS [   0.013s]       major        function_like_proc_macro_missing
        PASS [   0.016s]       major        function_missing
        PASS [   0.013s]       major        function_now_doc_hidden
        PASS [   0.010s]       major        function_parameter_count_changed
        PASS [   0.016s]       major        function_requires_different_const_generic_params
        PASS [   0.013s]       major        function_requires_different_generic_type_params
        PASS [   0.013s]       major        function_unsafe_added
        PASS [   0.013s]       major        inherent_associated_const_now_doc_hidden
        PASS [   0.027s]       major        inherent_associated_pub_const_missing
        PASS [   0.033s]       major        inherent_method_const_removed
        PASS [   0.025s]       major        inherent_method_missing
        PASS [   0.017s]       major        inherent_method_now_doc_hidden
        PASS [   0.030s]       major        inherent_method_unsafe_added
        PASS [   0.012s]       major        macro_no_longer_exported
        PASS [   0.009s]       major        macro_now_doc_hidden
        PASS [   0.020s]       major        method_parameter_count_changed
        PASS [   0.029s]       major        method_requires_different_const_generic_params
        PASS [   0.028s]       major        method_requires_different_generic_type_params
        PASS [   0.008s]       major        module_missing
        PASS [   0.006s]       major        non_exhaustive_struct_changed_type
        PASS [   0.013s]       major        partial_ord_enum_struct_variant_fields_reordered
        PASS [   0.011s]       major        partial_ord_enum_variants_reordered
        PASS [   0.024s]       major        partial_ord_struct_fields_reordered
        PASS [   0.007s]       major        proc_macro_now_doc_hidden
        PASS [   0.008s]       major        pub_module_level_const_missing
        PASS [   0.010s]       major        pub_module_level_const_now_doc_hidden
        PASS [   0.009s]       major        pub_static_missing
        PASS [   0.009s]       major        pub_static_mut_now_immutable
        PASS [   0.013s]       major        pub_static_now_doc_hidden
        PASS [   0.013s]       major        pub_static_now_mutable
        PASS [   0.014s]       major        repr_c_enum_struct_variant_fields_reordered
        PASS [   0.016s]       major        repr_c_plain_struct_fields_reordered
        PASS [   0.016s]       major        repr_c_removed
        PASS [   0.025s]       major        repr_packed_added
        PASS [   0.016s]       major        repr_packed_removed
        PASS [   0.036s]       major        sized_impl_removed
        PASS [   0.008s]       major        static_became_unsafe
        PASS [   0.012s]       major        struct_marked_non_exhaustive
        FAIL [   0.012s]       major        struct_missing
        PASS [   0.013s]       major        struct_now_doc_hidden
        PASS [   0.015s]       major        struct_pub_field_missing
        PASS [   0.019s]       major        struct_pub_field_now_doc_hidden
        PASS [   0.016s]       major        struct_repr_transparent_removed
        PASS [   0.025s]       major        struct_with_no_pub_fields_changed_type
        PASS [   0.014s]       major        struct_with_pub_fields_changed_type
        FAIL [   0.018s]       major        trait_added_supertrait
        PASS [   0.009s]       major        trait_allows_fewer_const_generic_params
        PASS [   0.011s]       major        trait_allows_fewer_generic_type_params
        PASS [   0.022s]       major        trait_associated_const_added
        PASS [   0.020s]       major        trait_associated_const_default_removed
        PASS [   0.010s]       major        trait_associated_const_now_doc_hidden
        PASS [   0.014s]       major        trait_associated_type_added
        PASS [   0.026s]       major        trait_associated_type_default_removed
        PASS [   0.014s]       major        trait_associated_type_now_doc_hidden
        PASS [   0.018s]       major        trait_method_added
        PASS [   0.016s]       major        trait_method_default_impl_removed
        PASS [   0.027s]       major        trait_method_missing
        PASS [   0.029s]       major        trait_method_now_doc_hidden
        PASS [   0.044s]       major        trait_method_parameter_count_changed
        PASS [   0.022s]       major        trait_method_requires_different_const_generic_params
        PASS [   0.017s]       major        trait_method_requires_different_generic_type_params
        PASS [   0.010s]       major        trait_method_unsafe_added
        PASS [   0.006s]       major        trait_method_unsafe_removed
        PASS [   0.009s]       major        trait_mismatched_generic_lifetimes
        PASS [   0.018s]       major        trait_missing
        PASS [   0.013s]       major        trait_newly_sealed
        FAIL [   0.012s]       major        trait_no_longer_dyn_compatible
        FAIL [   0.016s]       major        trait_now_doc_hidden
        PASS [   0.011s]       major        trait_removed_associated_constant
        PASS [   0.012s]       major        trait_removed_associated_type
        PASS [   0.013s]       major        trait_removed_supertrait
        PASS [   0.014s]       major        trait_requires_more_const_generic_params
        PASS [   0.032s]       major        trait_requires_more_generic_type_params
        PASS [   0.016s]       major        trait_unsafe_added
        PASS [   0.014s]       major        trait_unsafe_removed
        PASS [   0.015s]       major        tuple_struct_to_plain_struct
        PASS [   0.034s]       major        type_allows_fewer_const_generic_params
        PASS [   0.035s]       major        type_allows_fewer_generic_type_params
        PASS [   0.027s]       major        type_mismatched_generic_lifetimes
        PASS [   0.032s]       major        type_requires_more_const_generic_params
        PASS [   0.038s]       major        type_requires_more_generic_type_params
        PASS [   0.014s]       major        union_field_added_with_all_pub_fields
        PASS [   0.013s]       major        union_field_added_with_non_pub_fields
        PASS [   0.011s]       major        union_field_missing
        PASS [   0.009s]       major        union_missing
        PASS [   0.008s]       major        union_now_doc_hidden
        PASS [   0.008s]       major        union_pub_field_now_doc_hidden
        PASS [   0.010s]       major        unit_struct_changed_kind
     Checked [   0.139s] 128 checks: 123 pass, 5 fail, 0 warn, 36 skip

--- failure enum_no_repr_variant_discriminant_changed: enum variant had its discriminant change value ---

Description:
The enum's variant had its discriminant value change. This breaks downstream code that used its value via a numeric cast like `as isize`.
        ref: https://doc.rust-lang.org/reference/items/enumerations.html#assigning-discriminant-values
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/enum_no_repr_variant_discriminant_changed.ron

Failed in:
  variant DatabaseErrorKind::NotNullViolation 5 -> 6 in /home/weiznich/Documents/rust/diesel/diesel/src/result.rs:140
  variant DatabaseErrorKind::CheckViolation 6 -> 7 in /home/weiznich/Documents/rust/diesel/diesel/src/result.rs:143
  variant DatabaseErrorKind::ClosedConnection 7 -> 9 in /home/weiznich/Documents/rust/diesel/diesel/src/result.rs:152
  variant DatabaseErrorKind::Unknown 8 -> 10 in /home/weiznich/Documents/rust/diesel/diesel/src/result.rs:155

--- failure struct_missing: pub struct removed or renamed ---

Description:
A publicly-visible struct cannot be imported by its prior path. A `pub use` may have been removed, or the struct itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/struct_missing.ron

Failed in:
  struct diesel::pg::sql_types::Jsonb, previously in file /home/weiznich/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.2.10/src/pg/types/mod.rs:328

--- failure trait_added_supertrait: non-sealed trait added new supertraits ---

Description:
A non-sealed trait added one or more supertraits, which breaks downstream implementations of the trait
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#generic-bounds-tighten
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_added_supertrait.ron

Failed in:
  trait diesel::connection::Instrumentation gained Downcast in file /home/weiznich/Documents/rust/diesel/diesel/src/connection/instrumentation.rs:246

--- failure trait_no_longer_dyn_compatible: trait no longer dyn compatible ---

Description:
Trait is no longer dyn compatible, which breaks `dyn Trait` usage.
        ref: https://doc.rust-lang.org/stable/reference/items/traits.html#object-safety
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_no_longer_dyn_compatible.ron

Failed in:
  trait Foldable in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/fold.rs:4
  trait SingleValue in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/mod.rs:655
  trait SqlOrd in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/ord.rs:7
  trait SqlType in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/mod.rs:671

--- failure trait_now_doc_hidden: pub trait is now #[doc(hidden)] ---

Description:
A pub trait is now #[doc(hidden)], removing it from the crate's public API.
        ref: https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#hidden
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_now_doc_hidden.ron

Failed in:
  trait IntoArrayExpression in file /home/weiznich/Documents/rust/diesel/diesel/src/pg/expression/array.rs:74

     Summary semver requires new major version: 5 major and 0 minor checks failed

### Generated System Information

System information:
-------------------
#### Software version

cargo-semver-checks 0.41.0 (793f25c)

#### Operating system

- OS: Linux (Fedora Linux 42)
- Kernel: 6.14.5-300.fc42.x86_64

#### Command-line

```bash
/home/weiznich/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.86.0 (adf9b6ad1 2025-02-28)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu

Cargo build configuration:
--------------------------
[alias]
xtask = ["run", "--package", "xtask", "--"]

Please file an issue on GitHub reporting your bug.
Consider adding the diagnostic information above, either manually or automatically through the link below:

https://github.com/obi1kenobi/cargo-semver-checks/issues/new?labels=C-bug&template=3-bug-report.yml&sys-info=%23%23%23%23%20Software%20version%0A%0Acargo-semver-checks%200.41.0%20%28793f25c%29%0A%0A%23%23%23%23%20Operating%20system%0A%0A-%20OS%3A%20Linux%20%28Fedora%20Linux%2042%29%0A-%20Kernel%3A%206.14.5-300.fc42.x86_64%0A%0A%23%23%23%23%20Command-line%0A%0A%60%60%60bash%0A%2Fhome%2Fweiznich%2F.cargo%2Fbin%2Fcargo-semver-checks%20semver-checks%20--bugreport%20%0A%60%60%60%0A%0A%23%23%23%23%20cargo%20version%0A%0A%60%60%60%0A%3E%20cargo%20-V%20%0Acargo%201.86.0%20%28adf9b6ad1%202025-02-28%29%0A%60%60%60%0A%0A%23%23%23%23%20Compile%20time%20information%0A%0A-%20Profile%3A%20release%0A-%20Target%20triple%3A%20x86_64-unknown-linux-gnu%0A-%20Family%3A%20unix%0A-%20OS%3A%20linux%0A-%20Architecture%3A%20x86_64%0A-%20Pointer%20width%3A%2064%0A-%20Endian%3A%20little%0A-%20CPU%20features%3A%20fxsr%2Csse%2Csse2%0A-%20Host%3A%20x86_64-unknown-linux-gnu%0A%0A&build-config=%5Balias%5D%0Axtask%20%3D%20%5B%22run%22%2C%20%22--package%22%2C%20%22xtask%22%2C%20%22--%22%5D%0A

**Comment by obi1kenobi** [comment]
Yes, I agree this is a false-positive, and would still have been a false-positive even if everything was in one crate.

It should minimize to a case like this:
```rust
pub trait Example: 'static {}

pub trait Super {}

impl<T: 'static> Super for T {}
```
then change:
```diff
- pub trait Example: 'static {}
+ pub trait Example: Super + 'static {}
```
Anything that implemented `Example` previously will still implement it, since the new trait has a blanket impl over a subset (here, the same set) of bounds of `Example`.

I'd love to solve this, but realistically this is going to require a ton of work that is unlikely to happen anytime soon. Hopefully we'll get there eventually!
